pkg/plugins/runtime/gateway/listener_generator.go

```diff
@@ -1,21 +1,243 @@
+	"google.golang.org/protobuf/types/known/wrapperspb"
 
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core/policy"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
 	"github.com/kumahq/kuma/pkg/core/resources/manager"
-	"github.com/kumahq/kuma/pkg/core/xds"
+	"github.com/kumahq/kuma/pkg/core/resources/store"
+	core_xds "github.com/kumahq/kuma/pkg/core/xds"
 	xds_context "github.com/kumahq/kuma/pkg/xds/context"
+	"github.com/kumahq/kuma/pkg/xds/envoy"
+	envoy_listeners "github.com/kumahq/kuma/pkg/xds/envoy/listeners"
+	v3 "github.com/kumahq/kuma/pkg/xds/envoy/listeners/v3"
+	envoy_names "github.com/kumahq/kuma/pkg/xds/envoy/names"
 	"github.com/kumahq/kuma/pkg/xds/generator"
 )
 
-// ListenerGenerator generates Kuma gateway listeners. Not implemented.
+const DefaultConnectionBuffer = 32 * 1024
+const DefaultRequestHeadersTimeoutMsec = 500 * time.Millisecond
+const DefaultStreamIdleTimeoutMsec = 5 * time.Second
+const DefaultConcurrentStreams = 100
+const DefaultInitialStreamWindowSize = 64 * 1024
+const DefaultInitialConnectionWindowSize = 1024 * 1024
+const DefaultIdleTimeout = 5 * time.Minute
+
+type gatewayPolicyAdaptor struct {
+	*core_mesh.GatewayResource
+}
+
+func (g gatewayPolicyAdaptor) Selectors() []*mesh_proto.Selector {
+	return g.Sources()
+}
+
+func findMatchingGateway(m manager.ReadOnlyResourceManager, dp *core_mesh.DataplaneResource) *core_mesh.GatewayResource {
+	gatewayList := &core_mesh.GatewayResourceList{}
+
+	if err := m.List(context.Background(), gatewayList, store.ListByMesh(dp.Meta.GetMesh())); err != nil {
+		return nil
+	}
+
+	candidates := make([]policy.DataplanePolicy, len(gatewayList.Items))
+	for i, gw := range gatewayList.Items {
+		candidates[i] = gatewayPolicyAdaptor{gw}
+	}
+
+	if p := policy.SelectDataplanePolicy(dp, candidates); p != nil {
+		return p.(gatewayPolicyAdaptor).GatewayResource
+	}
+
+	return nil
+}
+
+var _ policy.DataplanePolicy = gatewayPolicyAdaptor{}
+
+// ListenerGenerator generates Kuma gateway listeners.
 type ListenerGenerator struct {
 	Resources manager.ReadOnlyResourceManager
 }
 
 var _ generator.ResourceGenerator = ListenerGenerator{}
 
-func (ListenerGenerator) Generate(ctx xds_context.Context, proxy *xds.Proxy) (*xds.ResourceSet, error) {
-	return nil, errors.New("not implemented")
+func (l ListenerGenerator) Generate(_ xds_context.Context, proxy *core_xds.Proxy) (*core_xds.ResourceSet, error) {
+	gw := findMatchingGateway(l.Resources, proxy.Dataplane)
+	if gw == nil {
+		log.V(1).Info("no matching gateway for dataplane",
+			"name", proxy.Dataplane.Meta.GetName(),
+			"mesh", proxy.Dataplane.Meta.GetMesh(),
+		)
+
+		return nil, nil
+	}
+
+	log.V(1).Info(fmt.Sprintf("matched gateway %q to dataplane %q",
+		gw.Meta.GetName(), proxy.Dataplane.Meta.GetName()))
+
+	// Multiple listener specifications can have the same port. If
+	// they are compatible, then we can collapse those specifications
+	// down to a single listener.
```

Comment:
could you provide an example of Gateway with many listeners on the same port? What is the use case?

Reply:
For example,

* 2 different hostnames on the same port
* 1 hostname that can be targeted by different route types (ie. different tag sets)